librz/analysis/arch/x86/x86_il.c

```diff
@@ -0,0 +1,3806 @@
+	} \
+	RzILOpEffect *count = SETL("_cnt", x86_il_get_op(1)); \
+	RzILOpEffect *masked = SETL("_masked", LOGAND(VARL("_cnt_mask"), VARL("_cnt"))); \
+	RzILOpEffect *temp_count = SETL("_tmp_cnt", MOD(VARL("_masked"), UN(cnt_size, size)));
+
+/**
+ * ROL
+ * Rotate left
+ * Encoding: MI, M1, MC
+ */
+IL_LIFTER(rol) {
+	ROX_MACRO();
+
+	RzILOpBool *cond = NON_ZERO(VARL("_tmp_cnt"));
+	RzILOpEffect *repeat = SETL("_tmp_cf", MSB(VARL("_dest")));
+	repeat = SEQ2(repeat, SETL("_dest", ADD(SHIFTL0(VARL("_dest"), U8(1)), x86_bool_to_bv(VARL("_tmp_cf"), BITS_PER_BYTE * size))));
+	repeat = SEQ2(repeat, SETL("_tmp_cnt", SUB(VARL("_tmp_cnt"), UN(cnt_size, 1))));
+
+	RzILOpBool *if_cond1 = NON_ZERO(VARL("_masked"));
+	RzILOpEffect *true_eff1 = SETG(EFLAGS(CF), LSB(VARL("_dest")));
+
+	RzILOpBool *if_cond2 = EQ(VARL("_masked"), UN(cnt_size, 1));
+	RzILOpEffect *true_eff2 = SETG(EFLAGS(OF), XOR(MSB(VARL("_dest")), VARG(EFLAGS(CF))));
+
+	RzILOpEffect *set_dest = x86_il_set_op(0, VARL("_dest"));
+	RzILOpEffect *ret = SEQ9(dest, count_mask, count, masked, temp_count, REPEAT(cond, repeat), BRANCH(if_cond1, true_eff1, NULL), BRANCH(if_cond2, true_eff2, NULL), set_dest);
+
+	return ret;
+}
+
+/**
+ * ROR
+ * Rotate right
+ * Encoding: MI, M1, MC
+ */
+IL_LIFTER(ror) {
+	ROX_MACRO();
+
+	RzILOpBool *cond = NON_ZERO(VARL("_tmp_cnt"));
+	RzILOpEffect *repeat = SETL("_tmp_cf", LSB(VARL("_dest")));
+	repeat = SEQ2(repeat, SETL("_dest", ADD(SHIFTR0(VARL("_dest"), U8(1)), SHIFTL0(x86_bool_to_bv(VARL("_tmp_cf"), BITS_PER_BYTE * size), U8(size)))));
+	repeat = SEQ2(repeat, SETL("_tmp_cnt", SUB(VARL("_tmp_cnt"), UN(cnt_size, 1))));
+
+	RzILOpBool *if_cond1 = NON_ZERO(VARL("_masked"));
+	RzILOpEffect *true_eff1 = SETG(EFLAGS(CF), MSB(VARL("_dest")));
+
+	RzILOpBool *if_cond2 = EQ(VARL("_masked"), UN(cnt_size, 1));
+	RzILOpEffect *true_eff2 = SETG(EFLAGS(OF), XOR(MSB(VARL("_dest")), MSB(SHIFTL0(VARL("_dest"), U8(1)))));
+
+	RzILOpEffect *set_dest = x86_il_set_op(0, VARL("_dest"));
+	RzILOpEffect *ret = SEQ9(dest, count_mask, count, masked, temp_count, REPEAT(cond, repeat), BRANCH(if_cond1, true_eff1, NULL), BRANCH(if_cond2, true_eff2, NULL), set_dest);
+
+	return ret;
+}
+
+#undef ROX_MACRO
+
+/**
+ * RET
+ * Return (near pointer)
+ * Encoding: ZO, I
+ * Most modern x86-32 and x86-64 programs use this return instruction
+ */
+IL_LIFTER(ret) {
+	PopHelper ph = x86_pop_helper(analysis->bits / BITS_PER_BYTE /* BYTES */);
+	RzILOpEffect *ret = ph.eff;
+	/* We can use RSP, and the correct stack pointer will be resolved depending on bitness */
+	ret = SEQ2(ret, x86_il_set_reg(X86_REG_RSP, ph.val));
+
+	if (ins->structure->op_count == 1) {
+		/* Immediate operand (Encondig: I) */
+		ret = SEQ2(ret, x86_il_set_reg(X86_REG_RSP, ADD(x86_il_get_reg(X86_REG_RSP), UN(analysis->bits, ins->structure->operands[0].imm))));
+	}
+
+	return ret;
+}
+
+// /**
+//  * RETF
+//  * Return far pointer
+//  * Encoding: ZO, I
+//  * Rarely found in modern programs
```

Comment:
In modern - yes. Not so rare for DOS programs which we should be able to analyze as well. Rizin is the RE framework, often RE needs to recover the logic from the ancient software..

Reply:
I really think it's the best to implement this instruction alongside tracetesting. For now I suggest it to stay commented. I would implement it when I tracetest it.